services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/hakemus/Hakemusalue.kt

```diff
@@ -60,17 +65,99 @@ data class KaivuilmoitusAlue(
             )
         }
     }
+
+    override fun listChanges(index: Int, other: Hakemusalue?): List<String>? {
+        if (other == null) return listOf("areas[$index]")
+        other as KaivuilmoitusAlue
+        val changes = mutableListOf<String>()
+        changes.addAll(checkChangesInTyoalueet(index, tyoalueet, other.tyoalueet))
+        checkChange(index, KaivuilmoitusAlue::name, other)?.let { changes.add(it) }
+        checkChange(index, KaivuilmoitusAlue::katuosoite, other)?.let { changes.add(it) }
+        checkChange(index, KaivuilmoitusAlue::tyonTarkoitukset, other)?.let { changes.add(it) }
+        checkChange(index, KaivuilmoitusAlue::meluhaitta, other)?.let { changes.add(it) }
+        checkChange(index, KaivuilmoitusAlue::polyhaitta, other)?.let { changes.add(it) }
+        checkChange(index, KaivuilmoitusAlue::tarinahaitta, other)?.let { changes.add(it) }
+        checkChange(index, KaivuilmoitusAlue::kaistahaitta, other)?.let { changes.add(it) }
+        checkChange(index, KaivuilmoitusAlue::kaistahaittojenPituus, other)?.let { changes.add(it) }
+        checkChange(index, KaivuilmoitusAlue::lisatiedot, other)?.let { changes.add(it) }
+        // haittojenhallintasuunnitelma changes do not affect the area itself, therefore we add the
+        // root area here if there are changes in other fields.
+        if (changes.isNotEmpty()) {
+            changes.add(0, "areas[$index]")
+        }
+        changes.addAll(
+            checkChangesInHaittojenhallintasuunnitelma(
+                index,
+                haittojenhallintasuunnitelma,
+                other.haittojenhallintasuunnitelma,
+            )
+        )
+        return if (changes.isNotEmpty()) changes else null
+    }
 }
 
 data class Tyoalue(
     val geometry: Polygon,
     val area: Double,
     val tormaystarkasteluTulos: TormaystarkasteluTulos?,
-)
+) {
+    fun listChanges(alueIndex: Int, tyoalueIndex: Int, other: Tyoalue?): List<String>? {
+        if (other == null) return listOf("areas[$alueIndex].tyoalueet[$tyoalueIndex]")
+        val changes = mutableListOf<String>()
+        checkChange(alueIndex, tyoalueIndex, Tyoalue::geometry, other)?.let { changes.add(it) }
+        return if (changes.isNotEmpty())
+            listOf("areas[$alueIndex].tyoalueet[$tyoalueIndex]") + changes
```

Comment:
Instead of taking the previous indexes as parameters, it would be cleaner to take the previous path. So this method doesn't need to assume the context in which it's run. Similar to how the `path` parameters are used in validations.

Reply:
Changed this as suggested.